schema/avro/union.go

```diff
@@ -0,0 +1,482 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package avro
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/conduitio/conduit-commons/opencdc"
+	"github.com/hamba/avro/v2"
+	"github.com/modern-go/reflect2"
+)
+
+// unionResolver provides hooks before marshaling and after unmarshaling a value
+// with an Avro schema, which make sure that values under the schema type Union
+// are in the correct shape (see https://github.com/hamba/avro#unions).
+// NB: It currently supports union types nested in maps, but not nested in
+// slices. For example, hooks will not work for values like []any{[]any{"foo"}}.
+type unionResolver struct {
+	mapUnionPaths   []path
+	arrayUnionPaths []path
+	nullUnionPaths  []path
+	resolver        *avro.TypeResolver
+}
+
+// newUnionResolver takes a schema and extracts the paths to all maps and arrays
+// with union types. With this information the resolver can traverse the values
+// in BeforeMarshal and AfterUnmarshal directly to the value that needs to be
+// substituted.
+func newUnionResolver(schema avro.Schema) unionResolver {
+	var mapUnionPaths []path
+	var arrayUnionPaths []path
+	var nullUnionPaths []path
+	// traverse the schema and extract paths to all maps and arrays with a union
+	// as the value type
+	traverseSchema(schema, func(p path) {
+		switch {
+		case isMapUnion(p[len(p)-1].schema):
+			// path points to a map with a union type, copy and store it
+			pCopy := make(path, len(p))
+			copy(pCopy, p)
+			mapUnionPaths = append(mapUnionPaths, pCopy)
+		case isArrayUnion(p[len(p)-1].schema):
+			// path points to an array with a union type, copy and store it
+			pCopy := make(path, len(p))
+			copy(pCopy, p)
+			arrayUnionPaths = append(arrayUnionPaths, pCopy)
+		case isNullUnion(p[len(p)-1].schema):
+			// path points to a null union, copy and store it
+			pCopy := make(path, len(p)-1)
+			copy(pCopy, p[:len(p)-1])
+			nullUnionPaths = append(nullUnionPaths, pCopy)
+		}
+	})
+	return unionResolver{
+		mapUnionPaths:   mapUnionPaths,
+		arrayUnionPaths: arrayUnionPaths,
+		nullUnionPaths:  nullUnionPaths,
+		resolver:        avro.NewTypeResolver(),
+	}
+}
+
+// AfterUnmarshal traverses the value using the schema and finds all values that
+// have the Avro type Union. Those values are unmarshaled into a map with a
+// single key that contains the name of the type
+// (e.g. map[string]any{"string":"foo"}). This function takes that map and
+// extracts the actual value from it (e.g. "foo").
```

Comment:
This comment is not quite clear to many. First this part:
```
// AfterUnmarshal traverses the value using the schema and finds all values that
// have the Avro type Union.
```
I guess what is meant is that this function traverses `val` (the input value) and finds all fields that are of the union type?

```
// Those values are unmarshaled into a map with a
// single key that contains the name of the type
// (e.g. map[string]any{"string":"foo"}). 
```

Is the following meant by the above: those fields (that are of the union type) are represented as maps in the input `val`? So this function extracts the actual values from those maps?

Reply:
Yeah you got it right. I'll try to rephrase it based on your suggestions.

Reply:
Reads great!